csharp/src/Ice/Reference.cs

```diff
@@ -1210,35 +1247,59 @@ internal async ValueTask<IRequestHandler> GetRequestHandlerAsync(CancellationTok
+        // Marshal the non-null reference.
         internal void Write(OutputStream ostr)
         {
             if (IsFixed)
             {
+                // TODO: should be true only for the 1.1 encoding once we add Fixed support in the 2.0 encoding
                 throw new NotSupportedException("cannot marshal a fixed proxy");
             }
 
-            Identity.IceWrite(ostr);
-            ostr.WriteFacet(Facet);
-            ostr.WriteByte((byte)InvocationMode);
-            ostr.WriteBool(false); // secure option, always false (not used)
-            ostr.WriteByte((byte)Protocol);
-            ostr.WriteByte(0);
-            ostr.WriteByte(Encoding.Major);
-            ostr.WriteByte(Encoding.Minor);
-
-            ostr.WriteSize(Endpoints.Count);
-            if (Endpoints.Count > 0)
+            if (ostr.Encoding == Encoding.V11)
             {
-                Debug.Assert(AdapterId.Length == 0);
-                foreach (Endpoint endpoint in Endpoints)
+                Identity.IceWrite(ostr);
+                var proxyData = new ProxyData11(Facet.Length > 0 ? new string[] { Facet } : Array.Empty<string>(),
+                                                InvocationMode,
+                                                secure: false,
+                                                Protocol,
+                                                protocolMinor: 0,
+                                                Encoding);
+                proxyData.IceWrite(ostr);
+                ostr.WriteSequence(Endpoints, (ostr, endpoint) => ostr.WriteEndpoint(endpoint));
+
+                if (Endpoints.Count == 0)
                 {
-                    ostr.WriteEndpoint(endpoint);
+                    ostr.WriteString(AdapterId);
+                }
+                else
+                {
+                    Debug.Assert(AdapterId.Length == 0);
                 }
             }
             else
             {
-                ostr.WriteString(AdapterId);
+                Debug.Assert(ostr.Encoding == Encoding.V20);
+
+                ostr.Write(Endpoints.Count > 0 ? ProxyKind.Direct : ProxyKind.Indirect);
+
+                // For non ice1 proxies, invocation mode is not marshaled so we "adjust" it to Twoway which gets
+                // converted to null below.
+                InvocationMode adjustedMode = Protocol == Protocol.Ice1 ? InvocationMode : InvocationMode.Twoway;
+
+                var proxyData = new ProxyData20(Identity,
+                                                Protocol == Protocol.Ice2 ? null : Protocol,
+                                                Encoding == Encoding.V20 ? null : Encoding,
+                                                AdapterId.Length > 0 ? new string[] { AdapterId } : null,
+                                                adjustedMode != InvocationMode.Twoway ? adjustedMode : null,
+                                                Facet.Length > 0 ? Facet : null);
```

Comment:
maybe simpler
```suggestion
                var proxyData = new ProxyData20(Identity,
                                                Protocol == Protocol.Ice2 ? null : Protocol,
                                                Encoding == Encoding.V20 ? null : Encoding,
                                                AdapterId.Length > 0 ? new string[] { AdapterId } : null,
                                                Protocol == Protocol.Ice2 ? null : InvocationMode,
                                                Facet.Length > 0 ? Facet : null);
```

Reply:
I want to write null when Protocol != Protocol.Ice1 or Protocol == Protocol.Ice1 and InvocationMode == InvocationMode.Twoway; otherwise, InvocationMode. Your version writes InvocationMode all the time when Protocol != Protocol.Ice2.